add design philosophy document (#87)
opensouls/community: docs/pages/design-philosophy.mdx

# Design Philosophy

> A core belief of OPEN SOULS is that LLMs are amazing reasoning machines, similar to the prefrontal cortex of the brain, or the CPU of a computer, but lack the _rest of the brain_. The Soul Engine is the tool models the rest of the mind: agency, memory, emotion, drive, goal setting, and more.

Reviewer: ```suggestion
A core belief of OPEN SOULS is that LLMs are amazing reasoning machines, similar to the prefrontal cortex of the brain, or the CPU of a computer, but lack the _rest of the brain_. The Soul Engine is the tool for modeling the rest of the mind: giving agency, memory, emotion, drive, goal setting, and more.
```

The Soul Engine optimizes for rich, soulful experiences. It is a developer tool and hosted service to create and deploy mind-blowing experiences with AI Souls.  It is a rich ecosystem of APIs for developers to craft and _host_ their AI Soul experiences.

As a team and community, we make decisions about what goes into and what stays out of the Soul Engine. This document expresses our main goals and reasoning.

## Simplicity

No magic. Designing a mind is an inherently complex task, but understanding individual components of the Soul Engine should be straightforward.

We prioritize simplicity of an individual API over reducing the number of APIs.

We abstract only after real-world use.

## Functional, Debuggable, Predictable

The interactions with the engine should be functional whenever possible and default towards immutable where the developer experience is not weakened. When you interact with an API from the Soul Engine, you can anticipate how it will work and your interactions are debuggable.

## Interaction over Accuracy

> We optimize for soul, not for chat bots.

Reviewer 2: ❤️

The goal of the Soul Engine is to allow developers to create the most unbelievably awesome experiences and model minds realistically. We will always trade off accuracy for personality, emotion, realism, and subjective quality of interaction.

## Familiarity

When possible, we re-use patterns already known to developers (React, for instance). We make the Soul Engine’s API clearer by re-using interfaces and elements developers are already familiar with.

## A Sculptor's Tool

> The Soul Engine is not a magic black box for AI soul creation. Instead, it allows developers to focus on the exact soul and experience they want to build. This enables rapid iteration and experimentation, making it fun. Soul developers gradually shape their experience, testing and debugging as they build and the soul engine should support them from first idea to production.

Reviewer: ```suggestion
The Soul Engine is not a magic black box for AI soul creation. Instead, it allows developers to focus on the exact soul and experience they want to build. This enables rapid iteration and experimentation, making it fun. Soul developers gradually shape their experience, testing and debugging as they build and the Soul Engine should support them from first idea to production.
```

## Reduce Friction

Every API added to the Soul Engine is designed to solve a specific problem encountered during years of building AI souls. When a developer encounters something hard or annoying to use, the Soul Engine should have a feature that makes it easier and/or more fun to accomplish the same goal.

## Inspired by Nature

> The inner workings of the soul engine are inspired by our best research into how the human mind functions.

Reviewer: ```suggestion
The inner workings of the Soul Engine are inspired by our best research into how the human mind functions.
```

When researching new APIs or styles of working, we analyze neuroscience, psychology, and intuitive self-understanding. These biological systems influence the design of the Soul Engine’s API.

## Extensible

> The Soul Engine interacts with your applications where they live, allowing you to build whatever you want, with an AI soul at the core (or periphery) of your project.

Reviewer 2: could talk more about the advantages of having a soul that can be run independently of the product - easier to debug, can be tested in isolation, can be tweaked by less technical people

these are obvious to us but not immediately apparent to new users

Reviewer: likewise I would comment here again about the goal to support applications from inception, to experimentation, to production scaling

## Built to Connect Human / Machine

> The Soul Engine and its interfaces are generally written TypeScript, because the Soul Engine empowers creators to connect their community directly to an AI soul, and those touch points tend to have a javascript runtime.

Reviewer 2: i'm not sure i completely understand what you mean here

Reviewer: I think the point here is something to the effect of (but slightly better worded)

We believe frontend developers will be the soul developers of tomorrow, because frontend engineering is the design of stateful systems that are closest touch to user interactions. Therefore have chosen the most popular frontend language to program in.
